omegaconf/listconfig.py

```diff
@@ -500,20 +499,26 @@ def __iter__(self) -> Iterator[Any]:
 
     class ListIterator(Iterator[Any]):
         def __init__(self, lst: Any, resolve: bool) -> None:
-            self.iter = iter(lst.__dict__["_content"])
             self.resolve = resolve
+            self.iterator = iter(lst.__dict__["_content"])
             self.index = 0
+            from .nodes import ValueNode
+
+            self.ValueNode = ValueNode
 
         def __next__(self) -> Any:
-            v = next(self.iter)
 
+            x = next(self.iterator)
             if self.resolve:
-                v = v._dereference_node()
+                x = x._dereference_node()
 
-            if v._is_missing():
+            if x._is_missing():
                 raise MissingMandatoryValue(f"Missing value at index {self.index}")
+
             self.index = self.index + 1
-            return _get_value(v)
+            if isinstance(x, self.ValueNode):
+                return x._value()
```

Comment:
This seems like a change in behavior in case `x` is a container that would be None / Missing / an interpolation. Is that intended? (and I guess we need a test since nothing broke)

Reply:
it is intended. I mentioned it in one of the comments (tested on Hydra as well).
will add tests covering it.

Reply:
4d3053773790bb95aef113ba1e1781f55e8fdf34, this was a little pandora's box. see commit comment.

Reply:
Regarding the new commit: ok

Regarding the change in behavior, looking into it more closely, I'm not super comfortable with it for two reasons:
* It now gives a way for users to end up getting access to a container that is None / missing through the public API (which wasn't really the case until now, except by explicitly instantiating one)
* my_listconfig[0] doesn't give the same object type as iter(next(my_listconfig)) (and this is inconsistent with the behavior of DictConfig)

I may have missed your comment on this topic => is there a strong motivation for this?

Reply:
> Regarding the new commit: ok
> 
> Regarding the change in behavior, looking into it more closely, I'm not super comfortable with it for two reasons:
> 
> * It now gives a way for users to end up getting access to a container that is None / missing through the public API (which wasn't really the case until now, except by explicitly instantiating one)

It does not.

> * my_listconfig[0] doesn't give the same object type as iter(next(my_listconfig)) (and this is inconsistent with the behavior of DictConfig)
> 

It does.

```python
In [17]: lst = ListConfig([None, ListConfig(None)])

In [18]: itr = iter(lst)

In [19]: repr(next(itr))
Out[19]: 'None'

In [20]: repr(next(itr))
Out[20]: 'None'

In [21]: repr(next(itr))
---------------------------------------------------------------------------
StopIteration                             Traceback (most recent call last)
<ipython-input-21-b5146a100c72> in <module>
----> 1 repr(next(itr))

~/dev/omegaconf/omegaconf/listconfig.py in __next__(self)
    508         def __next__(self) -> Any:
    509 
--> 510             x = next(self.iterator)
    511             if self.resolve:
    512                 x = x._dereference_node()

StopIteration: 
```

You are confusing `__iter__()` with the non public _iter_ex(resolve=False).

Reply:
As for DictConfig, I am looking at it now.

Reply:
> > Regarding the new commit: ok
> > Regarding the change in behavior, looking into it more closely, I'm not super comfortable with it for two reasons:
> > 
> > * It now gives a way for users to end up getting access to a container that is None / missing through the public API (which wasn't really the case until now, except by explicitly instantiating one)
> 
> It does not.

```python
lst = ListConfig([ListConfig(None)])
item = next(iter(lst))
assert isinstance(item, ListConfig) and item._value() is None
```

> 
> > * my_listconfig[0] doesn't give the same object type as iter(next(my_listconfig)) (and this is inconsistent with the behavior of DictConfig)
> 
> It does.
> 
> ```python
> In [17]: lst = ListConfig([None, ListConfig(None)])
> 
> In [18]: itr = iter(lst)
> 
> In [19]: repr(next(itr))
> Out[19]: 'None'
> 
> In [20]: repr(next(itr))
> Out[20]: 'None'

You're only looking at the string representation, but:

```python
lst = ListConfig([None, ListConfig(None)])
itr = iter(lst)
print(next(itr) is None)  # True
print(next(itr) is None)  # False
```

Reply:
Yeah, just saw that too.
will take a closer look.

Reply:
Sounds good